packages/auth/lib/hashSessionId.js

```diff
@@ -0,0 +1,7 @@
+module.exports = (id) => {
+  const crypto = require('crypto')
+  return crypto
+    .createHmac('sha256', process.env.SESSION_SECRET)
```

Comment:
don't use the SESSION_SECRET here. I could
- get my session ID from the dev tools
- get my HMACed session ID from the API
- start a bruteforce attach to get SESSION_SECRET
invent a independent ENV like SESSION_ID_HMAC and use checkENV to make sure it's set.

Reply:
@patte random thought: create a secret per user and save it in the db. We have similar needs for user specific secrets with display author. What do you think?

Reply:
I'm not sure on this. For the record: as you pointed out in https://github.com/orbiting/republik-backend/pull/16#issuecomment-344665606 there is not much more security gained from a secret per user than a global secret in regard to stealing it. If someone steals the env she also steals the DB credentials. But attacking the secrets from the outside gets much harder. And this is true here: Even if the attacker would break her salt she could not decypher other users session id. But this can just be mitigated by a, from the sid, independent `id2` on the sessions table, that is used for the interaction with the API. This ids are only visible to their owners and actions on it are protected by meOrHasRoles, so faking a HMAC is of no use.
Similarly a user specific salt could be used, but It would not help for display author, because we show comments to not-signed-in users, which don't have a session and therefor no persistent salt. And because the display author ids need to be stable over requests and websockets also a temporary/connection salt is hard to do.
Long story short, the advantage of a user specific salt is small. What do you guys think?
NOTE: [pgcrypto](https://www.postgresql.org/docs/current/static/pgcrypto.html#idm46428634902160). A 128b key should be enough

Reply:
Good reads: 
- https://benlog.com/2008/06/19/dont-hash-secrets/
- https://security.stackexchange.com/questions/29951/salted-hashes-vs-hmac

From what I read we could actually use PBKDF2 which uses HMAC several times internally and use the ip as a salt which would make it extremely expensive to do bruteforcing or rainbow table attacks against the system, but the only way how we can mitigate that attack vector pretty much completely is generating a UUID in the db (id2).

An HMAC incorporates additional secret data in the form of a key. This key is combined with the message in a much deeper way than a salt, and may be used in addition to a simple salt of the message. The end result is not vulnerable to the same attack that a simple hash with a secret salt would be, and so is a much stronger method for producing an authenticated hash.

**A key is not a salt (keys are secret, salts are not)**. But the unique characteristics of HMAC make it a reasonable building block for other functions, which is what happens in PBKDF2: that's a key derivation function, commonly subverted into password hashing (a role for which it appears to be adequate). PBKDF2 includes a salt, and internally invokes HMAC (several times).